include/seqan3/range/decorator/gap_decorator_anchor_set.hpp

```diff
@@ -0,0 +1,485 @@
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL KNUT REINERT OR THE FU BERLIN BE LIABLE
+// FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+// DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+// SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+// CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+// LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+// OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+// DAMAGE.
+//
+// ============================================================================
+
+/*!\file
+ * \author Marie Hoffmann <marie.hoffmann AT fu-berlin.de>
+ * \brief Contains a gap decorator to annotate sequences with gaps using a set.
+ */
+
+#pragma once
+
+#include <set>
+#include <type_traits>
+
+#include <range/v3/all.hpp>
+
+#include <seqan3/alphabet/concept.hpp>
+#include <seqan3/alphabet/gap/gapped.hpp>
+#include <seqan3/range/container/concept.hpp>
+#include <seqan3/range/detail/random_access_iterator.hpp>
+
+// ------------------------------------------------------------------
+// gap_decorator_anchor_set
+// ------------------------------------------------------------------
+
+namespace seqan3 {
+
+/*!\brief A gap decorator allows the annotation of sequences with gap symbols
+ * leaving the underlying sequence unmodified.
+ * \tparam inner_type The type of alphabet sequences that will be stored.
+ * \implements seqan3::aligned_sequence_concept.
+ * \ingroup decorator
+ *
+ * This class may be used whenever you want to store or compute an alignment.
+ * The underlying (ungapped sequence) remains unmodified, and is augmented by
+ * gap information. Iterating over a gap_decorator sequences behaves as if you
+ * had a rolled out, aligned sequence with alphabet and gap symbols.
+ *
+ * \details
+ * The gap_decorator_anchor_set is a modified anchor list approach - instead
+ * of storing tuples of anchor positions and gap lengths relative to the
+ * underlying sequence position, the anchor addresses are virtual and gap
+ * lengths accumulated from left to right, i.e. an anchor gap stores at the 2nd
+ * position all previous gap lengths including its own.
+ * This reduces the lookup time to log (k) with k
+ * being the number of continuous gaps (not gap symbols) and comes at the price
+ * of O(k) worst-case runtime for gap modifications - when inserting or erasing
+ * gaps the tailing gaps have to be updated by the resulting offset.
+ * The gap maintaining structure is a set that is given a gap comparing structure.
+ * Sets are implemented as red-black trees and perform random read operations in
+ * log (k) time. The anchor set approach provides a good trade-off when using
+ * both operator[] and gap insertion/erasure.
+ */
+template <typename inner_type>
+//!\cond
+    requires container_concept<inner_type>
```

Comment:
Yes, sounds right. Where is the docu for this? That I can check the features that a `std::ranges::RandomAccessRange` has to fulfill? If you have any idea which requirements of the `aligned_sequence_concept` I'm missing, I'd be glad to know them. The unit test for this concept is failing currently.

Reply:
http://docs.seqan.de/seqan/3.0.0-master-user/interfacestd_1_1ranges_1_1RandomAccessRange.html

but this is modelled e.g. by std::vector

Reply:
It would only tell that `myconcept` is not satisfied, but not which of its requirements out of the conditions list. I'll figure out by partial exclusion .